x/stakeibc/keeper/msg_server_liquid_stake_test.go

```diff
@@ -0,0 +1,215 @@
+	suite.CompareCoins(expectedUserAtomBalance, actualUserAtomBalance, "user ibc/uatom balance")
+	suite.CompareCoins(expectedModuleAtomBalance, actualModuleAtomBalance, "module ibc/uatom balance")
+
+	// Confirm deposit record adjustment
+	records := suite.App.RecordsKeeper.GetAllDepositRecord(suite.Ctx)
+	suite.Require().Len(records, 1, "number of deposit records")
+
+	expectedDepositRecordAmount := tc.initialState.depositRecordAmount + stakeAmount.Int64()
+	actualDepositRecordAmount := records[0].Amount
+	suite.Require().Equal(expectedDepositRecordAmount, actualDepositRecordAmount, "deposit record amount")
+}
+
+func (suite *KeeperTestSuite) TestLiquidStakeDifferentRedemptionRates() {
+	tc := suite.SetupLiquidStake()
+	user := tc.user
+	msg := tc.validMsg
+
+	// Loop over exchange rates: {0.2, 0.4, 0.6, ..., 2.0}
+	for i := -8; i <= 10; i += 2 {
+		redemptionDelta := sdk.NewDecWithPrec(1.0, 1).Mul(sdk.NewDec(int64(i))) // i = 2 => delta = 0.2
+		newRedemptionRate := sdk.NewDec(1.0).Add(redemptionDelta)
+		redemptionRateFloat := newRedemptionRate.MustFloat64()
+
+		// Update rate in host zone
+		hz := tc.initialState.hostZone
+		hz.RedemptionRate = newRedemptionRate
+		suite.App.StakeibcKeeper.SetHostZone(suite.Ctx, hz)
+
+		// Liquid stake for each balance and confirm stAtom minted
+		startingStAtomBalance := suite.App.BankKeeper.GetBalance(suite.Ctx, user.acc, "stuatom").Amount.Int64()
+		_, err := suite.msgServer.LiquidStake(sdk.WrapSDKContext(suite.Ctx), &msg)
+		suite.Require().NoError(err)
+		endingStAtomBalance := suite.App.BankKeeper.GetBalance(suite.Ctx, user.acc, "stuatom").Amount.Int64()
+		actualStAtomMinted := endingStAtomBalance - startingStAtomBalance
+
+		expectedStAtomMinted := int64(float64(msg.Amount) / redemptionRateFloat)
+		testDescription := fmt.Sprintf("st atom balance for redemption rate: %v", redemptionRateFloat)
+		suite.Require().Equal(expectedStAtomMinted, actualStAtomMinted, testDescription)
+	}
+}
+
+func (suite *KeeperTestSuite) TestLiquidStakeHostZoneNotFound() {
+	tc := suite.SetupLiquidStake()
+	// Update message with invalid denom
+	invalidMsg := tc.validMsg
+	invalidMsg.HostDenom = "ufakedenom"
+	_, err := suite.msgServer.LiquidStake(sdk.WrapSDKContext(suite.Ctx), &invalidMsg)
+
+	suite.Require().EqualError(err, "no host zone found for denom (ufakedenom): host zone not registered")
+}
+
+func (suite *KeeperTestSuite) TestLiquidStakeIbcCoinParseError() {
+	tc := suite.SetupLiquidStake()
+	// Update hostzone with denom that can't be parsed
+	badHostZone := tc.initialState.hostZone
+	badHostZone.IBCDenom = "ibc.0atom"
+	suite.App.StakeibcKeeper.SetHostZone(suite.Ctx, badHostZone)
+	_, err := suite.msgServer.LiquidStake(sdk.WrapSDKContext(suite.Ctx), &tc.validMsg)
+
+	badCoin := fmt.Sprintf("%d%s", tc.validMsg.Amount, badHostZone.IBCDenom)
+	suite.Require().EqualError(err, fmt.Sprintf("failed to parse coin (%s): invalid decimal coin expression: %s", badCoin, badCoin))
```

Comment:
where did you get the second argument to `EqualError`? Did you have to trigger the error to find this string?

Reply:
You can technically do it though tracing the code. You'd see that the parse coin [error](https://github.com/Stride-Labs/stride/pull/87/files#diff-7af1afb39d0c4ae4d8e39395c899dfcb7e81a99901dd0f56213173e1e7d029afR36) returns "`failed to parse coin (%s)`". And that it also wraps the error from an upstream function "`invalid decimal coin expression: %s`". But yeah in practice it's a little easier to run the test once to find the message and copy it. 

I think for most of these I would copy the first half of the error message from the `msg_liquid_stake` file (since that's the message we really want to test), and then I'd get the other error message (the upstream error) by triggering the test once and viewing the output